File: app/src/main/java/com/example/odm/garbagesorthelper/core/net/ObserverManager.java
package com.example.odm.garbagesorthelper.core.net;

import android.util.Log;

import com.orhanobut.logger.Logger;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

/**
 * Observer的继承类,优化了对各类网络异常的处理
 * @author: ODM
 * @date: 2019/8/17
 */
public abstract class ObserverManager<T >  implements Observer <T>{

    private static final String TAG = "ObserverManager";
    @Override
    public void onSubscribe(Disposable d) {
        Log.d(TAG, "onSubscribe: ");
    }

    @Override
    public void onError(Throwable e) {
        if (e instanceof Exception) {
            Logger.d("此错误的信息: "  + e.getMessage());
            onError(ThrowableHandler.handleThrowable(e));
        } else {
            onError(new HttpThrowable(HttpThrowable.UNKNOWN,"未知错误",e));
        }
    }

    @Override
    public void onComplete() {
    }

    /**
     * 强制实现，异常的处理
     * @param httpThrowable 网络异常
     */
    public abstract void onError(HttpThrowable httpThrowable);

    /**
     * 强制实现，对回调数据的处理
     */
    @Override
    public abstract void onNext(T t);
}

File: app/src/main/java/com/example/odm/garbagesorthelper/model/source/net/ObserverManager.kt
package com.example.odm.garbagesorthelper.model.source.net

import android.util.Log
import com.orhanobut.logger.Logger
import io.reactivex.Observer
import io.reactivex.disposables.Disposable

/**
 * Observer的继承类,优化了对各类网络异常的处理
 * @author: ODM
 * @date: 2019/8/17
 */
abstract class ObserverManager<T> : Observer<T> {
    override fun onSubscribe(d: Disposable) {
        Log.d(TAG, "onSubscribe: ")
    }

    override fun onError(e: Throwable) {
        if (e is Exception) {
            Logger.d("此错误的信息: " + e.message)
            onError(ThrowableHandler.handleThrowable(e))
        } else {
            onError(HttpThrowable(HttpThrowable.Companion.UNKNOWN, "未知错误", e))
        }
    }

    override fun onComplete() {}
    /**
     * 强制实现，异常的处理
     * @param httpThrowable 网络异常
     */
    abstract fun onError(httpThrowable: HttpThrowable)

    /**
     * 强制实现，对回调数据的处理
     */
    abstract override fun onNext(t: T)

    companion object {
        private const val TAG = "ObserverManager"
    }
}